src/redux/session.js

```diff
@@ -62,87 +61,59 @@ module.exports.setStatus = status => ({
+    if (
+        body.user &&
+        body.user.banned &&
+        banWhitelistPaths.indexOf(window.location.pathname) === -1
+    ) {
+        window.location = '/accounts/banned-response/';
+        return;
+    } else if (
+        body.flags &&
+        body.flags.must_complete_registration &&
+        window.location.pathname !== '/classes/complete_registration'
+    ) {
+        window.location = '/classes/complete_registration';
+        return;
+    } else if (
+        body.flags &&
+        body.flags.must_reset_password &&
+        !body.flags.must_complete_registration &&
+        window.location.pathname !== '/classes/student_password_reset/'
+    ) {
+        window.location = '/classes/student_password_reset/';
+        return;
+    }
+    dispatch(module.exports.setSession(body));
+    dispatch(module.exports.setStatus(module.exports.Status.FETCHED));
+
+    // get the permissions from the updated session
+    dispatch(permissionsActions.storePermissions(body.permissions));
+    if (typeof body.user !== 'undefined') {
+        dispatch(messageCountActions.getCount(body.user.username));
+    }
+    return;
+};
+
 module.exports.refreshSession = () => (dispatch => {
     dispatch(module.exports.setStatus(module.exports.Status.FETCHING));
-    api({
-        host: '',
-        uri: '/session/'
-    }, (err, body) => {
-        if (err) return dispatch(module.exports.setSessionError(err));
-        if (typeof body === 'undefined') return dispatch(module.exports.setSessionError('No session content'));
-        if (
-            body.user &&
-            body.user.banned &&
-            banWhitelistPaths.indexOf(window.location.pathname) === -1
-        ) {
-            window.location = '/accounts/banned-response/';
-            return;
-        } else if (
-            body.flags &&
-            body.flags.must_complete_registration &&
-            window.location.pathname !== '/classes/complete_registration'
-        ) {
-            window.location = '/classes/complete_registration';
-            return;
-        } else if (
-            body.flags &&
-            body.flags.must_reset_password &&
-            !body.flags.must_complete_registration &&
-            window.location.pathname !== '/classes/student_password_reset/'
-        ) {
-            window.location = '/classes/student_password_reset/';
-            return;
-        }
-        dispatch(module.exports.setSession(body));
-        dispatch(module.exports.setStatus(module.exports.Status.FETCHED));
-
-        // get the permissions from the updated session
-        dispatch(permissionsActions.storePermissions(body.permissions));
-        if (typeof body.user !== 'undefined') {
-            dispatch(messageCountActions.getCount(body.user.username));
-        }
-        return;
-    });
+    return new Promise((resolve, reject) => (
+        sessionLib.requestSessionOnce(resolve, reject).then(body => {
```

Comment:
It would read better to me if you just imported the functions you wanted to use and didn't refer to the library here.

Reply:
I find it clearer to know where they're coming from, but I'm happy to do it that way!